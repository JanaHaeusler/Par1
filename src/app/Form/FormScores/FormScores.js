import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import { v4 as uuid } from 'uuid'
import Button from '../../Button'
import { CancelIconDark, CheckIconLight } from '../../Icons/Icons'

FormScores.propTypes = {
  formInputs: PropTypes.object.isRequired,
  isSaveButtonShown: PropTypes.bool.isRequired,
  updateDirtyInputs: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  showErrorMessage: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleCancel: PropTypes.func.isRequired,
}

export default function FormScores({
  formInputs,
  isSaveButtonShown,
  updateDirtyInputs,
  handleChange,
  showErrorMessage,
  handleSubmit,
  handleCancel,
}) {
  return (
    <Form onSubmit={handleSubmit} data-testid="form">
      <Headline>Score</Headline>
      <ScoreOverview>
        <Legend>
          <span>Holes</span>
          {new Array(18).fill().map((_, index) => {
            const newId = uuid()
            return <span key={newId}>{index + 1}</span>
          })}
        </Legend>
        <ScoresAllPlayers>
          {formInputs.playersArray?.map((player, index) => {
            const playerName = player
            const key = `singlePlayerInputs${index + 1}`
            return (
              <ScoreSinglePlayer key={key}>
                <span>{playerName}</span>
                {new Array(18).fill().map((_, index) => {
                  const inputNumber = index + 1
                  const holeName = 'hole' + inputNumber
                  const key = `singleScoreInput${inputNumber}`
                  return (
                    <SingleScore key={key}>
                      <input
                        type="number"
                        name={holeName + '-' + playerName}
                        id={holeName + '-' + playerName}
                        aria-label={'score ' + holeName + '-' + playerName}
                        value={formInputs.scores[playerName][holeName]}
                        onChange={(event) => {
                          handleChange(event.target.name, event.target.value)
                        }}
                        onBlur={(event) => updateDirtyInputs(event.target.name)}
                      />
                    </SingleScore>
                  )
                })}
              </ScoreSinglePlayer>
            )
          })}
        </ScoresAllPlayers>
      </ScoreOverview>
      <Note>{showErrorMessage()}</Note>
      <ButtonWrapper>
        <Button
          type="button"
          onClick={handleCancel}
          iconComponent={<CancelIconDark />}
          text="Cancel"
          data-testid="button-cancel"
        />
        <Button
          main
          disabled={!isSaveButtonShown}
          iconComponent={<CheckIconLight />}
          text="Save"
          data-testid="button-save"
        />
      </ButtonWrapper>
    </Form>
  )
}

const Form = styled.form`
  margin: 10px 20px;
  padding: 10px;
  display: grid;
  place-items: center;
  border-radius: 25px;
  box-shadow: 0 0 10px var(--primary-medium);
  background-color: var(--text-light);
  color: var(--text-dark);
  font-size: 1rem;

  span {
    font-family: 'Raleway', sans-serif;
  }

  input::-webkit-outer-spin-button,
  input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  input[type='number'] {
    -moz-appearance: textfield;
  }

  Button {
    margin-top: 15px;
    margin-bottom: 5px;
  }
`
const Headline = styled.h3`
  text-align: center;
  text-transform: uppercase;
`
const ScoreOverview = styled.section`
  margin: 20px auto;
  display: grid;
  grid-template-columns: 1fr 3fr;
  gap: 20px;
  width: 90%;
`
const Legend = styled.div`
  padding-right: 20px;
  display: grid;
  grid-template-rows: repeat(19, 35px);
  align-items: center;
  border-right: 1px solid var(--separator);

  span {
    width: 100%;
    text-align: center;
  }
`
const ScoresAllPlayers = styled.div`
  display: flex;
  overflow-y: auto;
  scrollbar-width: none;

  &::-webkit-scrollbar {
    display: none;
  }
`
const ScoreSinglePlayer = styled.div`
  margin: 0 2px;
  padding-left: 1px;
  display: grid;
  grid-template-rows: repeat(19, 35px);
  align-items: center;
  min-width: 80px;
`
const SingleScore = styled.label`
  input {
    padding: 0;
    width: 100%;
    height: 100%;
    border-style: none;
    border-radius: 3px;
    background-color: var(--separator);
    color: var(--primary-dark);
    font-family: 'Montserrat', sans-serif;
  }
`
const Note = styled.span`
  margin: 5px 0 10px;
  font-size: 0.7rem;
  color: var(--text-dark);
`
const ButtonWrapper = styled.div`
  display: flex;
  justify-content: space-evenly;
  margin-bottom: 10px;
  width: 100%;
`
