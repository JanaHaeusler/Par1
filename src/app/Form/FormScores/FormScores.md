Form Scores

```jsx
<FormScores
  formInputs={{
    location: 'City Park',
    date: '2020-12-24',
    winner: 'Sarah',
    shots: '28',
    playersString: 'Sarah, Hannes',
    playersArray: ['Sarah', 'Hannes'],
    scores: {
      Sarah: {
        hole1: '',
        hole2: '',
        hole3: '',
        hole4: '',
        hole5: '',
        hole6: '',
        hole7: '',
        hole8: '',
        hole9: '',
        hole10: '',
        hole11: '',
        hole12: '',
        hole13: '',
        hole14: '',
        hole15: '',
        hole16: '',
        hole17: '',
        hole18: '',
      },
      Hannes: {
        hole1: '',
        hole2: '',
        hole3: '',
        hole4: '',
        hole5: '',
        hole6: '',
        hole7: '',
        hole8: '',
        hole9: '',
        hole10: '',
        hole11: '',
        hole12: '',
        hole13: '',
        hole14: '',
        hole15: '',
        hole16: '',
        hole17: '',
        hole18: '',
      },
    },
    _id: 1,
  }}
  isSaveButtonShown={false}
  updateDirtyInputs={() => {}}
  handleChange={() => {}}
  showErrorMessage={() => {}}
  handleSubmit={() => {}}
  handleCancel={() => {}}
/>
```
