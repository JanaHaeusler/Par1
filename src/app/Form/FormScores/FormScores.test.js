import { fireEvent, render } from '@testing-library/react'
import user from '@testing-library/user-event'
import FormScores from './FormScores'

const testProps = {
    formInputs: {
        location: 'City Park',
        date: '2020-12-24',
        winner: 'Sarah',
        shots: '28',
        playersString: 'Sarah, Hannes',
        playersArray: ['Sarah', 'Hannes'],
        scores: {
          Sarah: {
            hole1: '',
            hole2: '',
            hole3: '',
            hole4: '',
            hole5: '',
            hole6: '',
            hole7: '',
            hole8: '',
            hole9: '',
            hole10: '',
            hole11: '',
            hole12: '',
            hole13: '',
            hole14: '',
            hole15: '',
            hole16: '',
            hole17: '',
            hole18: '',
          },
          Hannes: {
            hole1: '',
            hole2: '',
            hole3: '',
            hole4: '',
            hole5: '',
            hole6: '',
            hole7: '',
            hole8: '',
            hole9: '',
            hole10: '',
            hole11: '',
            hole12: '',
            hole13: '',
            hole14: '',
            hole15: '',
            hole16: '',
            hole17: '',
            hole18: '',
          },
        },
        _id: '1',
      },
    isSaveButtonShown: false,
    updateDirtyInputs: jest.fn(),
    handleChange: jest.fn(),
    showErrorMessage: jest.fn(),
    handleSubmit: jest.fn(),
    handleCancel: jest.fn(),
}

describe('FormScores', () => {

    beforeEach(() => {
        Object.defineProperty(window, "localStorage", {
            value: {
            getItem: jest.fn(() => null),
            setItem: jest.fn(() => null)
            },
            writable: true
        })
    })
  
  it('renders snapshot correctly', () => {
    const { container } = render(<FormScores {...testProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('has two buttons', () => {
    const { getByText } = render(<FormScores {...testProps} />)
    expect(getByText('Save')).toBeInTheDocument()
    expect(getByText('Cancel')).toBeInTheDocument()
  })

  it('has disabled save button', () => {
    const { getByTestId} = render(<FormScores {...testProps} />)
    expect(getByTestId('button-save')).toBeDisabled()
  })

  it('has enabled save button', () => {
    const props = {...testProps, isSaveButtonShown: true}
    const { getByTestId} = render(<FormScores {...props} />)
    expect(getByTestId('button-save')).toBeEnabled()
  })

  it('calls updateDirtyInputsMock on blur event', () => {
    const updateDirtyInputsMock = jest.fn()
    const props = {...testProps, updateDirtyInputs: updateDirtyInputsMock}
    const { getByLabelText } = render(<FormScores {...props} />)
    const RandomTestInput = getByLabelText('score hole1-Sarah')
    fireEvent.blur(RandomTestInput)
    expect(updateDirtyInputsMock).toHaveBeenCalled()
  })

  it('calls handleChangeMock on change event', () => {
    const handleChangeMock = jest.fn()
    const props = {...testProps, handleChange: handleChangeMock}
    const { getByLabelText } = render(<FormScores {...props} />)
    const RandomTestInput = getByLabelText('score hole16-Hannes')
    fireEvent.change(RandomTestInput, { target: { value: '5' } })
    expect(handleChangeMock).toHaveBeenCalledTimes(1)
    expect(handleChangeMock).toHaveBeenCalledWith('hole16-Hannes', '5')
  })

  it('calls showErrorMessageMock', () => {
    const showErrorMessageMock = jest.fn()
    const props = {...testProps, showErrorMessage: showErrorMessageMock}
    render(<FormScores {...props} />)
    expect(showErrorMessageMock).toHaveBeenCalledTimes(1)
  })

  it('calls handleSubmitMock on submitting the form', () => {
    const handleSubmitMock = jest.fn()
    const props = {...testProps, isSaveButtonShown: true, handleSubmit: handleSubmitMock}
    const { getByTestId } = render(<FormScores {...props} />)
    const Form = getByTestId('form')
    fireEvent.submit(Form)
    expect(handleSubmitMock).toHaveBeenCalled()
  })

  it('calls handleCancelMock on clicking cancel button', () => {
    const handleCancelMock = jest.fn()
    const props = {...testProps, isSaveButtonShown: true, handleCancel: handleCancelMock}
    const { getByTestId } = render(<FormScores {...props} />)
    user.click(getByTestId('button-cancel'))
    expect(handleCancelMock).toHaveBeenCalled()
  })
})