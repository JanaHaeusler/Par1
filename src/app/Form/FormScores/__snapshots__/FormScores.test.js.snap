// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormScores renders snapshot correctly 1`] = `
.c8 {
  padding: 7px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  width: 85px;
  height: 34px;
  border-style: none;
  border-radius: 5px;
  box-shadow: 2px 2px 8px var(--secondary-medium-transparent);
  background-color: var(--light);
  color: var(--primary);
  font-family: 'Raleway',sans-serif;
  font-size: 1rem;
  font-weight: 550;
}

.c8:disabled {
  opacity: 50%;
}

.c10 {
  padding: 7px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  width: 85px;
  height: 34px;
  border-style: none;
  border-radius: 5px;
  box-shadow: 2px 2px 8px var(--secondary-medium-transparent);
  background-color: var(--primary);
  color: var(--light);
  font-family: 'Raleway',sans-serif;
  font-size: 1rem;
  font-weight: 550;
}

.c10:disabled {
  opacity: 50%;
}

.c9 {
  margin-right: 3px;
  stroke: var(--primary);
}

.c11 {
  margin-right: 3px;
  width: 25%;
  stroke: var(--light);
}

.c0 {
  margin: 10px 20px;
  padding: 10px;
  display: grid;
  place-items: center;
  border-radius: 25px;
  background-color: var(--light);
  font-size: 1rem;
}

.c0 input::-webkit-outer-spin-button,
.c0 input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c0 input[type='number'] {
  -moz-appearance: textfield;
}

.c0 Button {
  margin-top: 15px;
  margin-bottom: 5px;
}

.c1 {
  margin: 10px auto 20px;
  display: grid;
  grid-template-columns: 1fr 3fr;
  gap: 20px;
  max-width: 90%;
}

.c2 {
  padding-right: 20px;
  display: grid;
  grid-template-rows: repeat(19,35px);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-right: 1px solid var(--secondary-light);
}

.c2 span {
  text-align: center;
  font-weight: 550;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow-y: auto;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c3::-webkit-scrollbar {
  display: none;
}

.c4 {
  margin: 0 2px;
  display: grid;
  grid-template-rows: repeat(19,35px);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 80px;
}

.c4 span:first-child {
  font-weight: 550;
}

.c5 input {
  padding: 5px 10px 5px;
  width: 100%;
  height: 100%;
  border-style: none;
  border-radius: 3px;
  background-color: var(--secondary-light);
  color: var(--secondary-medium);
  font-family: 'Raleway',sans-serif;
  font-size: 1rem;
}

.c6 {
  margin-top: 5px;
  font-size: 0.7rem;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  margin-bottom: 10px;
  width: 100%;
}

<form
  class="c0"
  data-testid="form"
>
  <h3>
    Score
  </h3>
  <section
    class="c1"
  >
    <div
      class="c2"
    >
      <span>
        Holes
      </span>
      <span>
        1
      </span>
      <span>
        2
      </span>
      <span>
        3
      </span>
      <span>
        4
      </span>
      <span>
        5
      </span>
      <span>
        6
      </span>
      <span>
        7
      </span>
      <span>
        8
      </span>
      <span>
        9
      </span>
      <span>
        10
      </span>
      <span>
        11
      </span>
      <span>
        12
      </span>
      <span>
        13
      </span>
      <span>
        14
      </span>
      <span>
        15
      </span>
      <span>
        16
      </span>
      <span>
        17
      </span>
      <span>
        18
      </span>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <span>
          Sarah
        </span>
        <label
          class="c5"
        >
          <input
            aria-label="score hole1-Sarah"
            id="hole1-Sarah"
            name="hole1-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole2-Sarah"
            id="hole2-Sarah"
            name="hole2-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole3-Sarah"
            id="hole3-Sarah"
            name="hole3-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole4-Sarah"
            id="hole4-Sarah"
            name="hole4-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole5-Sarah"
            id="hole5-Sarah"
            name="hole5-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole6-Sarah"
            id="hole6-Sarah"
            name="hole6-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole7-Sarah"
            id="hole7-Sarah"
            name="hole7-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole8-Sarah"
            id="hole8-Sarah"
            name="hole8-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole9-Sarah"
            id="hole9-Sarah"
            name="hole9-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole10-Sarah"
            id="hole10-Sarah"
            name="hole10-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole11-Sarah"
            id="hole11-Sarah"
            name="hole11-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole12-Sarah"
            id="hole12-Sarah"
            name="hole12-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole13-Sarah"
            id="hole13-Sarah"
            name="hole13-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole14-Sarah"
            id="hole14-Sarah"
            name="hole14-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole15-Sarah"
            id="hole15-Sarah"
            name="hole15-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole16-Sarah"
            id="hole16-Sarah"
            name="hole16-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole17-Sarah"
            id="hole17-Sarah"
            name="hole17-Sarah"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole18-Sarah"
            id="hole18-Sarah"
            name="hole18-Sarah"
            type="number"
            value=""
          />
        </label>
      </div>
      <div
        class="c4"
      >
        <span>
          Hannes
        </span>
        <label
          class="c5"
        >
          <input
            aria-label="score hole1-Hannes"
            id="hole1-Hannes"
            name="hole1-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole2-Hannes"
            id="hole2-Hannes"
            name="hole2-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole3-Hannes"
            id="hole3-Hannes"
            name="hole3-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole4-Hannes"
            id="hole4-Hannes"
            name="hole4-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole5-Hannes"
            id="hole5-Hannes"
            name="hole5-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole6-Hannes"
            id="hole6-Hannes"
            name="hole6-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole7-Hannes"
            id="hole7-Hannes"
            name="hole7-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole8-Hannes"
            id="hole8-Hannes"
            name="hole8-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole9-Hannes"
            id="hole9-Hannes"
            name="hole9-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole10-Hannes"
            id="hole10-Hannes"
            name="hole10-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole11-Hannes"
            id="hole11-Hannes"
            name="hole11-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole12-Hannes"
            id="hole12-Hannes"
            name="hole12-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole13-Hannes"
            id="hole13-Hannes"
            name="hole13-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole14-Hannes"
            id="hole14-Hannes"
            name="hole14-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole15-Hannes"
            id="hole15-Hannes"
            name="hole15-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole16-Hannes"
            id="hole16-Hannes"
            name="hole16-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole17-Hannes"
            id="hole17-Hannes"
            name="hole17-Hannes"
            type="number"
            value=""
          />
        </label>
        <label
          class="c5"
        >
          <input
            aria-label="score hole18-Hannes"
            id="hole18-Hannes"
            name="hole18-Hannes"
            type="number"
            value=""
          />
        </label>
      </div>
    </div>
  </section>
  <span
    class="c6"
  />
  <div
    class="c7"
  >
    <button
      class="c8"
      data-testid="button-cancel"
      type="button"
    >
      <svg
        class="c9"
      >
        cross.svg
      </svg>
      Cancel
    </button>
    <button
      class="c10"
      data-testid="button-save"
      disabled=""
    >
      <svg
        class="c11"
      >
        check.svg
      </svg>
      Save
    </button>
  </div>
</form>
`;
